drivers/dac/ad3552r/ad3552r.c

```diff
@@ -0,0 +1,1370 @@
+{
+	uint16_t code;
+	uint8_t addr;
+
+	addr = _get_code_reg_addr(ch, 1, 0);
+	if (desc->ch_data[ch].fast_en)
+		code = val & AD3552R_MASK_DAC_12B;
+	else
+		code = val;
+
+	return ad3552r_write_reg(desc, addr, code);
+}
+
+static int32_t _ad3552r_get_code_value(struct ad3552r_desc *desc,
+				       uint8_t ch,
+				       uint16_t *val)
+{
+	int32_t err;
+	uint8_t addr;
+
+	addr = _get_code_reg_addr(ch, 1, 0);
+	err = ad3552r_read_reg(desc, addr, val);
+	if (IS_ERR_VALUE(err))
+		return err;
+
+	return SUCCESS;
+}
+
+
+static void ad3552r_get_custom_range(struct ad3552r_desc *dac, uint8_t i,
+				     int32_t *v_min, int32_t *v_max)
+{
+	int64_t vref, tmp, common, offset, gn, gp;
+	/*
+	 * From datasheet formula (In Volts):
+	 *	Vmin = 2.5 + [(GainN + Offset / 1024) * 2.5 * Rfb * 1.03]
+	 *	Vmax = 2.5 - [(GainP + Offset / 1024) * 2.5 * Rfb * 1.03]
+	 * Calculus are converted to milivolts
+	 */
+	vref = 2500;
+	/* 2.5 * 1.03 * 1000 (To mV) */
+	common = 2575 * dac->ch_data[i].rfb;
+	offset = dac->ch_data[i].gain_offset;
+
+	gn = gains_scaling_table[dac->ch_data[i].n];
+	tmp = (1024 * gn + AD3552R_GAIN_SCALE * offset) * common;
+	tmp = tmp / (1024  * AD3552R_GAIN_SCALE);
+	*v_max = vref + tmp;
+
+	gp = gains_scaling_table[dac->ch_data[i].p];
+	tmp = (1024 * gp - AD3552R_GAIN_SCALE * offset) * common;
+	tmp = tmp / (1024 * AD3552R_GAIN_SCALE);
+	*v_min = vref - tmp;
+}
+
+static void ad3552r_calc_gain_and_offset(struct ad3552r_desc *dac, uint8_t ch)
+{
+	int32_t idx, v_max, v_min, span;
+	int64_t tmp, rem;
+
+	if (dac->ch_data[ch].range_override) {
+		ad3552r_get_custom_range(dac, ch, &v_min, &v_max);
+	} else {
+		/* Normal range */
+		idx = dac->ch_data[ch].range;
+		if (dac->chip_id == AD3542R_ID) {
+			v_min = ad3542r_ch_ranges[idx][0];
+			v_max = ad3542r_ch_ranges[idx][1];
+		} else {
+			v_min = ad3552r_ch_ranges[idx][0];
+			v_max = ad3552r_ch_ranges[idx][1];
+		}
+	}
+
+	/*
+	 * From datasheet formula:
+	 *	Vout = Span * (D / 65536) + Vmin
+	 * Converted to scale and offset:
+	 *	Scale = Span / 65536
+	 *	Offset = 65536 * Vmin / Span
+	 *
+	 * Reminders are in micros in order to be printed as
+	 * IIO_VAL_INT_PLUS_MICRO
+	 */
+	span = v_max - v_min;
+	rem = span % 65536;
+	dac->ch_data[ch].scale_int = span / 65536;
+	/* Do operations in microvolts */
+	dac->ch_data[ch].scale_dec = DIV_ROUND_CLOSEST(rem * 1000000, 65536);
+
+	tmp = v_min * 65536;
+	rem = tmp % span;
+	dac->ch_data[ch].offset_int = tmp / span;
+	dac->ch_data[ch].offset_dec = (rem * 1000000) / span;
+}
+
+static int32_t _ad3552r_set_gain_value(struct ad3552r_desc *desc,
+				       enum ad3552r_ch_attributes attr,
+				       uint8_t ch,
+				       uint16_t val)
+{
+	uint16_t reg_mask, reg;
+	int32_t err;
+
+	err = ad3552r_read_reg(desc,  AD3552R_REG_ADDR_CH_GAIN(ch), &reg);
+	if (IS_ERR_VALUE(err))
+		return err;
+
+	switch (attr) {
+	case AD3552R_CH_GAIN_OFFSET:
+		desc->ch_data[ch].offset = val;
+		err = ad3552r_write_reg(desc,  AD3552R_MASK_CH_OFFSET_BITS_0_7,
```

Comment:
Do we need register address here and not mask?

Reply:
ad3552r_write_reg is used so  the address register is needed.